### PR TITLE
Add indexer, checker and metrics to archive

### PR DESCRIPTION
**Pull request by tongzhi1998**
Add several additional features to monad-archive
1. tx indexer: key = tx_hash, value = Tx, Receipt, Trace, Header Subset
2. archive-checker: check the consistency of different S3 buckets, write errors and post to grafana 
3. index-checker: check if the program indexes everything correctly, write errors and post to grafana

### CODE REVIEW

File: monad-archive/src/archive_reader.rs [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    dynamodb: DynamoDBArchive,
+}
+
+impl ArchiveReader {
+    pub async fn new(
+        bucket: String,
+        table: String,
+        region: Option<String>,
+        concurrency: usize,
+        metrics: &Metrics,
```
Comment (eerkaijun) [inline monad-archive/src/archive_reader.rs]: `Option<Metrics>`

File: monad-archive/src/bin/monad-archive-checker/main.rs [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+#[tokio::main]
+async fn main() -> Result<()> {
+    tracing_subscriber::fmt().with_max_level(Level::INFO).init();
+
+    let args = cli::Cli::parse();
+    let metrics = Metrics::new(
+        args.otel_endpoint,
+        "monad-archiver",
+        Duration::from_secs(15),
+    )?;
```
Comment (eerkaijun) [inline monad-archive/src/bin/monad-archive-checker/main.rs]: maybe
```
let metrics = match args.otel_endpoint { 
    Some(otel_endpoint) => { Some(Metrics::new(...)) },
    None => None
}
```

File: monad-archive/src/dynamodb.rs [hunk 3]
```diff
@@ -0,0 +1,386 @@
...
+#[derive(
+    Debug, Clone, PartialEq, Eq, Default, Serialize, Deserialize, RlpEncodable, RlpDecodable,
+)]
+#[rlp(trailing)]
+pub struct HeaderSubset {
+    pub block_hash: BlockHash,
+    pub block_number: u64,
+    pub tx_index: u64,
+    pub gas_used: U256,
+    pub base_fee_per_gas: Option<U128>,
```
Comment (eerkaijun) [inline monad-archive/src/dynamodb.rs]: sorry my bad didn't raise this earlier, can `gas_used` be `u64` and `base_fee_per_gas` be `Option<u64>`?

File: monad-archive/src/dynamodb.rs [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+    async fn index_block(
+        &self,
+        block: Block,
+        traces: Vec<Vec<u8>>,
+        receipts: Vec<ReceiptWithBloom>,
+    ) -> Result<()> {
+        let mut requests = Vec::new();
+        let block_number = block.number;
+        let block_hash = block.hash_slow();
+        let base_fee_per_gas = block.base_fee_per_gas.map(U128::from);
```
Comment (eerkaijun) [inline monad-archive/src/dynamodb.rs]: as per above comment, easier to not convert to reth type here: 
```
let base_fee_per_gas = block.base_fee_per_gas;
```

File: monad-archive/src/dynamodb.rs [hunk 5]
```diff
@@ -0,0 +1,386 @@
...
+            let mut attribute_map = HashMap::new();
+            attribute_map.insert("tx_hash".to_owned(), AttributeValue::S(hex::encode(hash)));
+
+            // header subset
+            let mut rlp_header_subset = Vec::with_capacity(512);
+            HeaderSubset {
+                block_hash,
+                block_number,
+                tx_index: idx as u64,
+                gas_used: U256::from(gas_used_vec[idx]),
```
Comment (eerkaijun) [inline monad-archive/src/dynamodb.rs]: similarly, no need to convert to reth type

File: monad-archive/src/metrics.rs [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    pub meter: Meter,
+}
+
+impl Metrics {
+    pub fn new(
+        otel_endpoint: Option<impl AsRef<str>>,
+        service_name: impl Into<String>,
+        interval: Duration,
+    ) -> Result<Metrics> {
+        let provider = build_otel_meter_provider(otel_endpoint, service_name.into(), interval)?;
```
Comment (michael-yxchen) [inline monad-archive/src/metrics.rs:26]: @tongzhi1998 you had some issues with pushing metrics. Does keeping `provider` from dropping fix it?

File: docker/archive/Dockerfile [hunk 7]
```diff
@@ -70,21 +70,60 @@ RUN --mount=type=cache,target=${CARGO_ROOT}/registry    \
...
+RUN strip \
+    archive-checker \
+    *.so 
+
+RUN strip \
+    indexer \
+    *.so 
+
+RUN strip \
+    index-checker \
```
Comment (michael-yxchen) [inline docker/archive/Dockerfile]: combine them into one strip command

File: docker/archive/Dockerfile [hunk 8]
```diff
@@ -70,21 +70,60 @@ RUN --mount=type=cache,target=${CARGO_ROOT}/registry    \
...
+    CC=gcc-13 CXX=g++-13 cargo build --release --bin monad-indexer && \
+    mv target/release/monad-indexer indexer && \
+    cp `ls -Lt $(find target/release | grep -e "libtriedb_driver.so") | awk -F/ '!seen[$NF]++'` . 
+
+RUN --mount=type=cache,target=${CARGO_ROOT}/registry    \
+    --mount=type=cache,target=${CARGO_ROOT}/git          \
+    --mount=type=cache,target=/usr/src/monad-bft/target \
+    ASMFLAGS="-march=haswell" CFLAGS="-march=haswell" CXXFLAGS="-march=haswell" \
+    CC=gcc-13 CXX=g++-13 cargo build --release --bin monad-index-checker && \
+    mv target/release/monad-index-checker index-checker && \
```
Comment (michael-yxchen) [inline docker/archive/Dockerfile]: Combine into one cargo build

File: monad-archive/Cargo.toml [hunk 9]
```diff
@@ -5,26 +5,50 @@ version = "0.1.0"
...
 monad-triedb-utils = { path = "../monad-triedb-utils" }
 
-alloy-primitives = { workspace = true }
 alloy-rlp = { workspace = true }
 aws-config = { workspace = true }
+aws-sdk-dynamodb = { workspace = true }
 aws-sdk-s3 = { workspace = true }
 bytes = { workspace = true }
 clap = { workspace = true, features = ["derive"] }
+chrono = { workspace = true }
```
Comment (michael-yxchen) [inline monad-archive/Cargo.toml:34]: nit: sort

**Pull request merged**